src/psyclone/psyir/symbols/generic_interface_symbol.py

```diff
@@ -0,0 +1,178 @@
+        if not values:
+            raise ValueError("A GenericInterfaceSymbol requires a list of "
+                             "RoutineSymbols but none were provided.")
+        if not isinstance(values, list):
+            raise TypeError(f"A GenericInterfaceSymbol requires a list of "
+                            f"tuples describing its member routines but got: "
+                            f"'{values}'")
+        self._routines = []
+        for item in values:
+            if not isinstance(item, tuple):
+                raise TypeError(
+                    f"A GenericInterfaceSymbol ('{self.name}') requires a "
+                    f"list of tuples but got: '{type(item).__name__}'")
+            if (len(item) != 2 or not isinstance(item[0], RoutineSymbol) or
+                    not isinstance(item[1], bool)):
+                raise TypeError(
+                    f"Each tuple used to define a routine within the "
+                    f"GenericInterfaceSymbol '{self.name}' must consist of a "
+                    f"RoutineSymbol and a bool but got: "
+                    f"{[type(rt).__name__ for rt in item]}")
+            self._routines.append(self.RoutineInfo(item[0], item[1]))
+
+    @property
+    def container_routines(self):
+        '''
+        :returns: those routines that are defined in a Container.
+        :rtype: list[:py:class:`psyclone.psyir.symbols.RoutineSymbol`]
+        '''
+        result = []
+        for value in self._routines:
+            if value.from_container:
+                result.append(value.symbol)
+        return result
+
+    @property
+    def external_routines(self):
+        '''
+        :returns: those routines that are external procedures.
+        :rtype: list[:py:class:`psyclone.psyir.symbols.RoutineSymbol`]
+        '''
+        result = []
+        for value in self._routines:
+            if not value.from_container:
+                result.append(value.symbol)
+        return result
+
+    def replace(self, existing_sym, new_sym):
+        '''
+        Replace the specified RoutineSymbol with the new RoutineSymbol.
+
+        The 'from_container' property of the entry is preserved.
+
+        :param existing_sym: the RoutineSymbol to replace.
+        :type existing_sym: :py:class:`psyclone.psyir.symbols.RoutineSymbol`
+        :param new_sym: the new RoutineSymbol.
+        :type new_sym: :py:class:`psyclone.psyir.symbols.RoutineSymbol`
+
+        :raises ValueError: if the supplied RoutineSymbol is not referenced
+                            by this interface.
+        '''
+        for idx, value in enumerate(self._routines):
+            if value.symbol is existing_sym:
+                self._routines[idx] = self.RoutineInfo(new_sym,
+                                                       value.from_container)
+                return
+        raise ValueError(f"RoutineSymbol '{existing_sym.name}' not found in "
+                         f"GenericInterfaceSymbol '{self.name}'")
```

Comment:
I have probably missed it, but where is this method used/needed?

Reply:
It's used in the adjoint construction after a routine has been renamed (which means we need a new RoutineSymbol).

Reply:
See the other comment, if that works, I think this should be removed to let the symbol table manage symbol changes.